### DISCUSSION TITLE
Concept of a strict component

### DISCUSSION THREAD

**Discussion by mattbha** [wrabit/django-cotton#190] [General]
_Sorry for the discussion spam. Just an idea I had. Food for thought. I think implementation would be straightforward._

`<c-strict />`

💡 If this was added to a cotton component, the template would raise an exception on rendering if a variable was failed to be provided to the context. Ie

cotton/danger.html
-----
```
<c-strict />
<div class="alert alert-danger">
	watch out {{ username }} you cannot click {{ object.id }}!
</div>
```

This would be used as normal

```
<c-danger :username=request.user.username :object=invoice />
```

But due to the presence of `<c-strict />` if it was invoked WITHOUT a keyword argument, it would raise an exception

```
<c-danger :username=request.user.username />
```

^ Would raise exception missing argument `object`

**Comment by mattbha** [comment]
we could even force defining all used variables in c-strict (if provided). as in the above example, we could make it so you'd have to use

`<c-strict username object />`

otherwise the template would fail entirely, with "strict component uses undefined context variables". this could be a strong way of "forcing" the user to define all variables used in one place (making it easy for the consumer to tell whats used) and also to prevent them from forgetting to pass a variable

this is especially useful as the components get more complex, and less useful for small components

**Reply by wrabit** [reply]
Would it be mainly for catching mistakes during development? Would something like this suffice? https://stackoverflow.com/a/35554103/841591

Not ruling anything out here and I hear people often desiring better checking of template vars, just haven't needed this myself so far.. For critical areas I am writing tests.

**Reply by mattbha** [reply]
while it would aid mainly in development it would also be for long term maintainability. it'd be like defining your parameters at the top of a python function, and refusing to run the function if they aren't defined. the main win would be having the developer not need to search the html for {{ variables }}.

its useful sometimes to have the empty variables be a valid path, so the middleware wouldn't work. I would want this behavior on a component by component case, not everywhere.

could even call it `<c-assert username object />`
